addons/godot-next/references/bit_flag.gd

```diff
@@ -0,0 +1,95 @@
+# BitFlag
+# author: xdgamestudios
+# license: MIT
+# description: A class that allows to convert enums into bit flag enums and
+#              provides abstraction API to work with them.
+# todo: Implement additional features
+# deps:
+# - ClassType
+# usage:
+# - Initial setup:
+#   enum my_enum { a, b, c }
+#
+# - Creating:
+#     bf = BitFlag.new(my_enum) # Creates a new BitFlag given an enum.
+# - Flag access:
+#     bf.<name> = true/false # Allows to set the current bit
+# - Set/Put flag:
+#     bf.put(bf.a) # Enables the flag 'a'. Another BitFlag can be provided.
+# - Clear flag:
+#     bf.clear(bf.b) # Disables the flag 'b'. Another BitFlag can be provided.
+# - Toggle flag:
+#     bf.toggle(bf.b) # Inverts the flag 'b'. Another BitFlag can be provided.
+# - Check flag:
+#     bf.check(bf.b) # Checks flag 'b'. Another BitFlag can be provided.
+# - Get active:
+#     bf.get_active() # Returns an array of active flags be name.
+tool
+extends Reference
+class_name BitFlag
+
+##### CONSTANTS #####
+
+var SELF_PATH = "res://addons/godot-next/references/bit_flag.gd"
+
+##### PROPERTIES #####
+
+var _enum: Dictionary = {}
+var _flags: int = 0
+
+#warning-ignore:unused_class_variable
+var _class_type: ClassType = ClassType.new(SELF_PATH)
+
+##### NOTIFICATIONS #####
+
+func _init(p_enum: Dictionary):
+	for a_key in p_enum:
+		_enum[a_key] = 1 << p_enum[a_key]
+
+func _get(p_property: String) -> int:
+	if _enum.has(p_property):
+		return _enum[p_property]
+	return 0
+
+func _set(p_property: String, p_value: bool) -> bool:
+	if _enum.has(p_property):
+		if p_value:
+			_flags |= _enum[p_property]
+		else:
+			_flags &= _enum[p_property]
+		return true
+	return false
+
+##### PUBLIC METHODS #####
+
+func put(p_value) -> int:
+	var flags: int = p_value._flags if _class_type.is_type(p_value) else p_value
```

Comment:
You only need to use the whole `SELF_PATH` thing if you are attempting to access the current class from a static method (because the static method can't call any function local to the script.....like `get_script()`).

Rather than checking `if _class_type.is_type(p_value)`, you can instead just do `if get_script() == p_value.get_script()`. 

Also, '_flags |= flags` is a bit misleading I think. The value given to us is expected to be a single collection of bits, with only one of the values being edited. Maybe calling it `bit` instead? And then doing `_flags |= bit`? idk. Maybe that's a bad idea. What do you think? Technically you are handling a full byte integer, so...

Reply:
Also, add a method that allows me to complete re-assign a new value to the BitFlags instance.